### PR TITLE
made error sync and handle redis io error

### LINKED ISSUE #64: Make errors sync and improve handling of redis errors
* drop connections only on IO errors
* errors are no `Sync`

### PR DESCRIPTION
**Pull request by chridou**
closes #64

### CODE REVIEW

**Comment by chridou** [comment]
:+1:

**Comment by chridou** [comment]
:+1:

**Comment by chridou** [comment]
:+1:

File: src/pools/pool_internal/managed.rs [hunk 1]
```diff
@@ -24,34 +24,46 @@ pub(crate) struct Managed<T: Poolable> {
...
+    pub fn connection_mut(&mut self) -> Option<&mut T> {
+        self.connection.as_mut()
+    }
+
+    pub fn connection(&self) -> Option<&T> {
+        self.connection.as_ref()
+    }
+
+    /// Takes the connection which will prevent it from returning to the pool
+    #[allow(dead_code)]
```
Comment (panicbit) [inline src/pools/pool_internal/managed.rs]: cfg(test)

File: src/pools/pool_internal/test.rs [hunk 2]
```diff
@@ -145,8 +145,8 @@ fn checkout_one() {
     thread::sleep(Duration::from_millis(10));
 
     let v = runtime.block_on(async {
-        let check_out = check_out_fut(&pool, Wait).await.unwrap();
-        check_out.value.unwrap()
+        let mut check_out = check_out_fut(&pool, Wait).await.unwrap();
```
Comment (panicbit) [inline src/pools/pool_internal/test.rs]: => check_out_managed

File: src/pools/pool_internal/test.rs [hunk 3]
```diff
@@ -199,15 +199,15 @@ fn checkout_twice_with_delay_factory_with_one_not_reusable() {
...
+        let mut checked_out_managed = check_out_fut(&pool, Wait).await.unwrap();
+        checked_out_managed.take_connection().unwrap()
     });
 
     assert_eq!(v, 0);
 
     let v = runtime.block_on(async {
-        let check_out = check_out_fut(&pool, Wait).await.unwrap();
-        check_out.value.unwrap()
+        let mut check_out = check_out_fut(&pool, Wait).await.unwrap();
```
Comment (panicbit) [inline src/pools/pool_internal/test.rs]: => managed

**Comment by panicbit** [comment]
👍

**Pull request merged**